src/ripple/app/misc/NetworkOPs.cpp

```diff
@@ -2921,6 +2926,29 @@ InfoSub::pointer NetworkOPsImp::addRpcSub (
     return rspEntry;
 }
 
+bool NetworkOPsImp::tryRemoveRpcSub (std::string const& strUrl)
+{
+    ScopedLockType sl (mSubLock);
+    auto pInfo = findRpcSub(strUrl);
+
+    // check to see if any of the stream maps still hold a weak reference to
+    // this entry before removing
+    if ( pInfo &&
+         mSubLedger.find(pInfo->getSeq()) == mSubLedger.end() &&
```

Comment:
I agree with @bachase that this `if` looks fragile during maintenance.

When I saw this I really wanted to be able to look at whether a `std::shared_ptr` has one or more associated `std::weak_ptr`.  The library implementation knows this, but you can't get to it through the API because the information is easy to abuse.  Bummer.

I have a couple of other ideas for data structure changes that might make this less fragile.  The first thought is to simply move the list of `map`s that you're operating on closer to the definitions of the `map`s themselves.  Here's an example implementation of the idea: https://github.com/scottschurr/rippled/commit/780bac8bc0892dc435eec4958fcd38f47cc0063b

The downside of this idea is that we still have double book keeping, it's just that the required duplication is closer to the original definitions.

A slightly more radical thought is to use a `std::array<SubMapType>` to hold all of the associated `map`s.  Then we index the `array` with an `enum` so the indices have good names.  Here's an example implementation of the idea: https://github.com/scottschurr/rippled/commit/24f72bbcc24b6bdf47c00dd924274d6e599e1c2f.

I see two downsides to this choice.  The indexing leads to longer names, so we have more characters in the file (for those who are counting).  Also it's possible, during maintenance, for the `array` to become too small.  I hoped that clang would give me a warning if that happened, but no such luck.  Too bad, because both the array size and the index being used are known at compile time.

Don't feel obligated to use either idea.  I suggest them mostly to loosen up our thinking.